File: backend/src/main/java/com/gxs/bank/service/NotificationService.java
package com.gxs.bank.service;

import com.gxs.bank.model.Notification;
import com.gxs.bank.model.User;
import com.gxs.bank.repository.NotificationRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class NotificationService {
    private final NotificationRepository notificationRepository;

    public void createNotification(User user, String title, String message, Notification.Type type) {
        Notification notification = Notification.builder()
            .user(user)
            .title(title)
            .message(message)
            .type(type)
            .isRead(false)
            .build();
        notificationRepository.save(notification);
    }

    public List<Notification> getNotifications(UUID userId) {
        return notificationRepository.findByUserIdOrderByCreatedAtDesc(userId);
    }

    public long getUnreadCount(UUID userId) {
        return notificationRepository.countByUserIdAndIsReadFalse(userId);
    }

    public void markAsRead(UUID id) {
        notificationRepository.findById(id).ifPresent(n -> {
            n.setIsRead(true);
            notificationRepository.save(n);
        });
    }
}

File: backend-python/com/gxs/bank/service/NotificationService.py
from com.gxs.bank.model.Notification import Notification
from com.gxs.bank.repository.NotificationRepository import NotificationRepository


class NotificationService:
    def __init__(self, db):
        self.db = db
        self.notificationRepository = NotificationRepository(db)

    def createNotification(self, user, title: str, message: str, notificationType, auto_commit: bool = True):
        notification = Notification(
            user=user,
            userId=user.id,
            title=title,
            message=message,
            type=notificationType,
            isRead=False,
        )
        self.notificationRepository.save(notification)
        if auto_commit:
            self.db.commit()
            self.db.refresh(notification)
        return notification

    def getNotifications(self, userId: str):
        return self.notificationRepository.findByUserIdOrderByCreatedAtDesc(userId)

    def getUnreadCount(self, userId: str):
        return self.notificationRepository.countByUserIdAndIsReadFalse(userId)

    def markAsRead(self, notificationId: str):
        notification = self.notificationRepository.findById(notificationId)
        if notification is not None:
            notification.isRead = True
            self.db.commit()
